### ISSUE TITLE
trustedOrigins should be empty by default

### ISSUE THREAD

**Issue by kjetilk** [solid/node-solid-server#1090]
Community Leader @timbl decided that the default for trusted origins should contain no domains, POD providers should set that explicit in their configs.

We need to reflect that in the defaults. It also involves rewriting a test that now relies on that default.

**Comment by timbl** [comment]
Part of the logic for this is that a trusted domain is quite a centralized thing to build in, and while there will be a few domains probably which are very widely trusted, we don't want people to think that is part of the architecture. Another is the "secure by default" maxim.  Better to have a system more secure by default and have the system administrator explicitly  add the insecurity like a trusted domain.

**Comment by melvincarvalho** [comment]
:+1: 

> 20191131_Weekly Community Support Meeting

>    Decide on practical interim solution for solidproject.org domain and hosting, website upkeep, licensing body, Solid spec maintenance and upkeep:

https://github.com/solid/community/blob/master/community-support-agenda-and-minutes.md

I see there was a meeting around the solidproject.org domain in December.  I was wondering if any conclusions made in this meeting?

**Comment by melvincarvalho** [comment]
> @megoth Hey, just saw https://github.com/solid/node-solid-server/issues/1090#issuecomment-461887300, a bit weird place to bring up the question? It's a valid question, just thought the placement a bit weird =p

Thanks for raising this concern.  I put the question on this thread as the other one was closed.  Happy to hear suggestions of where you think it should go.

On a more constructive note, perhaps we could update the wiki on the current thought process.  I wasnt at the december meeting but last time this was discussed I think Tim favoured the forum to be a .org but solid doesnt have a .org right now, so it leads to the question of what to do with community sites short / medium term.  Or I wonder if eventually we'll move everything onto solid?

**Comment by kjetilk** [comment]
@melvincarvalho Since the document is in the community repo, I think it is better discussed there.

**Comment by melvincarvalho** [comment]
@kjetilk good thought, we could definitely raise an issue there, tho I suggest just someone who was at the meeting to update the wiki would suffice.  

There is a part relevant to node solid server tho.

https://github.com/solid/node-solid-server/pull/1077/files#diff-753c564b1569d68651fd4edd8ef2073cR15

Namely your pull request to add app.solidproject.org as a trusted domain to node solid server.  It would be good to have a context and throught process behind this PR.  Namely, to understand what exactly app.solidproject.org is, or what it is supposed to be.

**Comment by kjetilk** [comment]
Basically, the origin-based app authorization is a pain, and the question was what we could do to alleviate that pain. We had a face2face meeting around such topics, I don't remember exactly who was there, but I was there and @timbl and @RubenVerborgh discussed it at some length, and we resolved that beside implementing it like it is now is in acl-check.js, we could perhaps provide an origin for the community that most would trust by default, and that people could submit their apps to. It is clearly not a good solution, but since that is what we figured at the time, I made that PR. I don't have much opinion on it myself, we need to come up with better mechanisms. It is configurable for POD providers though, so now, the decision is rather to have them do it.

How the solidproject.org domain is managed is however a topic that is best suited over in the community repo. :-)

**Comment by melvincarvalho** [comment]
@kjetilk thanks for the clarification.  A few things still remain unclear.

What should pod providers put in the [trustedOrigin](https://github.com/solid/node-solid-server/pull/1077/files#diff-753c564b1569d68651fd4edd8ef2073cR15) config field for node solid server v5?

https://app.solidforum.org doesnt have anything there, so what is supposed to be there in theory.  Is it a place where we put copies of all the solid apps?

If so, should be put live versions of apps onto : https://github.com/solid/solid-apps ?

What should the format look like?  Do we want RDF meta data, something like a DOAP index?  Maybe an array of DOAPs that render into web pages.

Is the plan still to use the solid apps ontology?

https://github.com/solid/vocab/blob/master/solid-app.ttl

Questions, questions!

**Comment by kjetilk** [comment]
Questions, questions, indeed! :-)

Actually, my best answer is that I don't know, and I there isn't much concrete in the plans, I think, it was mostly just to future-proof things since the pain is easy to see. To me, the important part was to have a configurable array of trusted origins in v5, not what is actually in there.

Others may want to chime in with more, and this is a thread we need to take up, but we need a different forum, because I will need to close this issue once the corresponding PR (#1091) is merged.

**Comment by melvincarvalho** [comment]
@kjetilk it's still not clear what to put in this field.  Are there any docs for it?

**Comment by kjetilk** [comment]
> @kjetilk it's still not clear what to put in this field. Are there any docs for it?

You can put any origins all your users should trust. I suppose there aren't any hosts that are suitable to be used like that yet, so it is mainly future proofing. I can see that we should mention it in the README.

**Comment by melvincarvalho** [comment]
@kjetilk docs would be good.

I've opened this issue, so that we can collect apps in the solid area : 

https://github.com/solid/solid-apps/issues/25

The idea is to take a copy of existing apps and host them oh gh-pages.

Then we can give people live demos, and maybe play around with trusting origins.

**Comment by kjetilk** [comment]
I opened #1095 :-)

**Comment by melvincarvalho** [comment]
Thanks.  BTW I'm already working with trusted origins in the browser and payments with the [joule extension](https://github.com/wbobeirne/joule-extension) for lightning network.

They have a system that is working well.  No trusted origins by default, but users can white list apps (known as lapps) as they use them.  It's working really well with real world payments so far.

![image](https://user-images.githubusercontent.com/65864/52541940-fd560f80-2d9a-11e9-9a22-6adb48b7cad6.png)

**Comment by melvincarvalho** [comment]
The UI dialog : 

![image](https://user-images.githubusercontent.com/65864/52542002-c03e4d00-2d9b-11e9-8115-d26885aed95b.png)

**Comment by melvincarvalho** [comment]
So there's an old phrase in the decentralized web, 

> "No matter how decentralized you make a system, centralization creeps in through the back door".  

We need to be weary about potentially un-redecentralizting the web, by inserting trusted third parties, which is part of the model that solid is pushing back against.  I get a sense of deja vu.  For example, OIDC was introduced as having a trusted third party, to improve the UX, and it has become a source of confusion in node solid server.

If the default trusted origin is empty, are inrupt going to add app.solidproject.org to the trusted origins field manually?  And if so, will other providers need to do the same in order to provide a consistent UX?    

We to be clear about what trusted origins are, and how people are going to start to configure them.

**Comment by kjetilk** [comment]
> So there's an old phrase in the decentralized web,
> 
> > "No matter how decentralized you make a system, centralization creeps in through the back door".
> 
> We need to be weary about potentially un-redecentralizting the web, 

Absolutely. Indeed, there is a lot that indicates that decentralization  is an unstable situation. One reason for the introduction of `trustedOrigin` as a feature is that it is needed to indicate that you should trust your own origin. 

However, as interesting as this discussion is, it is not the place to discuss it further in this issue. This is just an issue to follow up on Tim's decision that this array should be empty by default.

**Comment by melvincarvalho** [comment]
> decentralization is an unstable situation

I suspect that's going to be a rather tough sell, here!

There are still unanswered questions about trusted origins.

**Comment by kjetilk** [comment]
> There are still unanswered questions about trusted origins.

Are you opposed to that the array of trusted origins should be empty by default?

**Comment by melvincarvalho** [comment]
> Are you opposed to that the array of trusted origins should be empty by default?

You havent explained the trade offs or the motivation for your pull request.

> Basically, the origin-based app authorization is a pain, and the question was what we could do to alleviate that pain.

You have said this will cause pain.  But not said what pain that will be.  

> have the system administrator explicitly add the insecurity like a trusted domain

System administrators dont know what to do.  If we are all going to have to set the defaults individually, can you at least explain what you or inrupt is going to add here.

We want people on different pods, say, to be able to chat to each other.

**Comment by megoth** [comment]
> > Are you opposed to that the array of trusted origins should be empty by default?
> 
> You havent explained the trade offs or the motivation for your pull request.
> 

He said that right in the OP, and @timbl even embellished on it:

> Part of the logic for this is that a trusted domain is quite a centralized thing to build in, and while there will be a few domains probably which are very widely trusted, we don't want people to think that is part of the architecture. Another is the "secure by default" maxim. Better to have a system more secure by default and have the system administrator explicitly add the insecurity like a trusted domain.

@melvincarvalho what would be your approach to this?

**Comment by melvincarvalho** [comment]
@megoth Approach :  

Design up front.  Document what trusted origins are, and how they affect apps.  What edge cases (eg CSRF) do they mitigate.  What are the pain points, one for users, one for pod owners.

Possible approaches for short and medium term relief, and what are the trade offs.  Any insertion of a specific domain into the code base needs to be covered transparently, and explaining what it is, and what will be on that domain.  

Minutes for decisions such as these should be kept so we know who asked for what.  I know it sounds like a lot of effort but trust is hard to gain, and easy to lose.  

What is needed now is to understand what should be put in this array, how it affects security, UX and interop.  While the default is empty, pod admins need to have a conversation about how they will configure this field, and what it implies.

One lesson we learnt from OIDC, is that when we inserted trusted third parties for auth, it got quite complicated fast, and that has lead to a lot of confusion.  It would be good to nip confusion in the bud with trusted origins.

Just my 2 Satoshis!

**Comment by megoth** [comment]
Those are good suggestions, but they will take long time; maybe we could accept this solution for now, and start another issue, with the approach you suggest?

Or do you think it's better to leave this issue for now?

**Comment by melvincarvalho** [comment]
Design etc. in another thread.

One this thread, it would be helpful to know how Inrupt will configure this.

**Comment by megoth** [comment]
The way @timbl wants us to do it, I reckon.

**Comment by megoth** [comment]
(hopefully he has time to look at this issue soon)

**Comment by kjetilk** [comment]
First of all, please understand that nothing is ever worked on in NSS without there being an issue here first. Everything is in the open. Always. However, to have progress, we also need other high-bandwidth venues to discuss, and interested parties need to be active here to influence. This thread was started in #738 based on high bandwidth discussions face-to-face, and minutes were, IIRC posted to gitter and cited in that issue. If you had concerns, those concerns would have been on topic on that issue.

In https://github.com/solid/solid-permissions/issues/26 you agreed to the proposed design of #1077. I simply do not understand your motivation for coming back now saying that this is not transparent and have not been up for debate, and we have not documented the decisions and the designs, when it is right there, and you engaged and agreed with the direction half a year ago. I believe that we cannot just run around in circles over and over again over topics like this, it is extremely destructive for our progress.

I have stated several times over that I have no strong opinions on what goes in that array. I have also stated several times over that we don't know what to put in that array. I have stated several times over that it is for future proofing. Perhaps it will never be used, as we may come up with a much better design. Moreover, with the closure of #1077 , we know what is *not* going in there by default. 

Also, I have responded to that even though it is OT.

What I do have a strong opinion about is how we should manage issues. The topic of this issue is whether the default should be empty, and only that. That is a topic that has already been decided by the Community Leader, who has the role of making such decision. Note, also, that the Community Leader is the author of the software component that introduced the `trustedOrigins` array in the current software, as we decided to discontinue the old permissions module. 

Moreover, it is easy to go and look at the actual patch. It replaces an intentionally invalid origin with an empty one. That is the only topic of this issue. Are there any tradeoffs between replacing an invalid origin intended for pre-release testing with an empty one? No, there isn't. But that is the topic of this issue, and only that.

I do in fact agree with the need to further document design and architectural decisions, in less volatile environs than gitter. However, we have nothing to build on. I am also reluctant to comment on that further since it is highly off topic for this issue.

**Comment by melvincarvalho** [comment]
@kjetilk thanks for the response.  A pointer to the minutes would be appreciated.

**Comment by melvincarvalho** [comment]
@kjetilk I do thank you for the useful pointers going back to August.  I'll note they were absent from the PR and would have provided a useful contest.  Digging them out here has been helpful.

However, as an inrupt employee your PR inserted the domain app.solidproject.org into the reference code base as the single default trusted origin for node solid server.  As far as I know that .org is soley owned by inrupt not the "we" implied by https://github.com/solid/node-solid-server/issues/738 which suggests the community.  

Polite requests clarification on this domain remain unanswered and were labeled, as off-topic.  There was apparently a community meeting about it, but the minutes are missing.

You seem to be saying that putting app.solidproject.org was perhaps not fully your idea.  It is reasonable for the community to ask for some kind of audit trail surrounding that decision.  If there was face2face around this where minutes exist, it is better just to post a link to the minutes.  

I have not sanctioned putting app.solidproject.org in as a trusted domain.  I would be open do doing so, but would need to know what it is.  Tim has eloquently described why this could be problematic.  It seems a double edged sword.

I know you have good intentions, and you work hard on solid.  But as I say, trust is easier to lose than it is to gain.  You could have simply diffused the situation as @megoth appeared to have done quite skillfully.  

What I suggest is that, in future, when you raise a PR that has a previous context, to add a link to that.  And when there is a decision made in a meeting, link to any minutes.  Casual members of the solid community are not going to be able to guess this.

We are now left as pod administrators to work out what to put in that field.  This will take some coordination and affects 1000s of users.

**Comment by melvincarvalho** [comment]
BTW  Dont mind this issue to be closed, but some items will benefit from follow up, such as what pod admins will do in a practical sense for v5.

I do think the right decisions were made.  And I realize I'm being super cautious here.  But I think it's an attitude that will pay off over time.  

Feel free to close, but let's try and make sure get  on top of trusted origins, before they get on top of us! :)
